Import Libraries

In [11]:
import pandas as pd
import numpy as np

Create the Load Function

In [13]:
def load_data(file_path):
    try:
        df = pd.read_csv(file_path, encoding='latin1')
        print("Dataset Loaded Successfully")
        return df

    except Exception as e:
        print("Error Loading File:", e)
        return None

Load the CSV File

In [15]:
file_path = "/Users/drashtisatani/Desktop/DA_python_Assessment/sales_data_sample.csv"

df = load_data(file_path)

Dataset Loaded Successfully


Check the Data

In [16]:
df.head()

,ORDERNUMBER,QUANTITYORDERED,PRICEEACH,ORDERLINENUMBER,SALES,ORDERDATE,STATUS,QTR_ID,MONTH_ID,YEAR_ID,...,ADDRESSLINE1,ADDRESSLINE2,CITY,STATE,POSTALCODE,COUNTRY,TERRITORY,CONTACTLASTNAME,CONTACTFIRSTNAME,DEALSIZE
0,10107,30,95.70,2,2871.00,2/24/2003 0:00,Shipped,1,2,2003,...,897 Long Airport Avenue,NaN,NYC,NY,10022,USA,NaN,Yu,Kwai,Small
1,10121,34,81.35,5,2765.90,5/7/2003 0:00,Shipped,2,5,2003,...,59 rue de l'Abbaye,NaN,Reims,NaN,51100,France,EMEA,Henriot,Paul,Small
2,10134,41,94.74,2,3884.34,7/1/2003 0:00,Shipped,3,7,2003,...,27 rue du Colonel Pierre Avia,NaN,Paris,NaN,75508,France,EMEA,Da Cunha,Daniel,Medium
3,10145,45,83.26,6,3746.70,8/25/2003 0:00,Shipped,3,8,2003,...,78934 Hillside Dr.,NaN,Pasadena,CA,90003,USA,NaN,Young,Julie,Medium
4,10159,49,100.00,14,5205.27,10/10/2003 0:00,Shipped,4,10,2003,...,7734 Strong St.,NaN,San Francisco,CA,NaN,USA,NaN,Brown,Julie,Medium


Clean the PRODUCTLINE Column

In [17]:
df["PRODUCTLINE"] = (
    df["PRODUCTLINE"]
    .astype(str)
    .str.strip()
    .str.title()
)

print("PRODUCTLINE cleaned successfully")

PRODUCTLINE cleaned successfully


Clean and Convert SALES Column

In [18]:
df["SALES"] = (
    df["SALES"]
    .astype(str)
    .str.replace(r"[^0-9.]", "", regex=True)
)

df["SALES"] = pd.to_numeric(df["SALES"], errors="coerce")

print("SALES column converted to float")

SALES column converted to float


Check Missing Values

In [19]:
print("Missing Values Report")
print("-" * 30)

print(df.isnull().sum())

Missing Values Report
------------------------------
ORDERNUMBER            0
QUANTITYORDERED        0
PRICEEACH              0
ORDERLINENUMBER        0
SALES                  0
ORDERDATE              0
STATUS                 0
QTR_ID                 0
MONTH_ID               0
YEAR_ID                0
PRODUCTLINE            0
MSRP                   0
PRODUCTCODE            0
CUSTOMERNAME           0
PHONE                  0
ADDRESSLINE1           0
ADDRESSLINE2        2521
CITY                   0
STATE               1486
POSTALCODE            76
COUNTRY                0
TERRITORY           1074
CONTACTLASTNAME        0
CONTACTFIRSTNAME       0
DEALSIZE               0
dtype: int64


Check Data Types

In [20]:
print(df.dtypes)

ORDERNUMBER           int64
QUANTITYORDERED       int64
PRICEEACH           float64
ORDERLINENUMBER       int64
SALES               float64
ORDERDATE            object
STATUS               object
QTR_ID                int64
MONTH_ID              int64
YEAR_ID               int64
PRODUCTLINE          object
MSRP                  int64
PRODUCTCODE          object
CUSTOMERNAME         object
PHONE                object
ADDRESSLINE1         object
ADDRESSLINE2         object
CITY                 object
STATE                object
POSTALCODE           object
COUNTRY              object
TERRITORY            object
CONTACTLASTNAME      object
CONTACTFIRSTNAME     object
DEALSIZE             object
dtype: object


Create Revenue Summary

In [21]:
summary = (
    df.groupby("PRODUCTLINE")["SALES"]
    .sum()
    .sort_values(ascending=False)
)

summary

PRODUCTLINE
Classic Cars        3919615.66
Vintage Cars        1903150.84
Motorcycles         1166388.34
Trucks And Buses    1127789.84
Planes               975003.57
Ships                714437.13
Trains               226243.47
Name: SALES, dtype: float64

Categorize Products

In [22]:
print("PRODUCT PERFORMANCE")
print("-" * 40)

for product, sales in summary.items():

    if sales > 100000:
        print(f"{product} --> High Revenue Product (Total Sales = {sales:,.2f})")
    else:
        print(f"{product} --> Low Revenue Product (Total Sales = {sales:,.2f})")

PRODUCT PERFORMANCE
----------------------------------------
Classic Cars --> High Revenue Product (Total Sales = 3,919,615.66)
Vintage Cars --> High Revenue Product (Total Sales = 1,903,150.84)
Motorcycles --> High Revenue Product (Total Sales = 1,166,388.34)
Trucks And Buses --> High Revenue Product (Total Sales = 1,127,789.84)
Planes --> High Revenue Product (Total Sales = 975,003.57)
Ships --> High Revenue Product (Total Sales = 714,437.13)
Trains --> High Revenue Product (Total Sales = 226,243.47)


Detect Outliers

In [23]:
Q1 = df["SALES"].quantile(0.25)
Q3 = df["SALES"].quantile(0.75)

IQR = Q3 - Q1

lower_bound = Q1 - (1.5 * IQR)
upper_bound = Q3 + (1.5 * IQR)

outliers = df[
    (df["SALES"] < lower_bound) |
    (df["SALES"] > upper_bound)
]

print("Number of Outliers:", len(outliers))

Number of Outliers: 81


Final Automated Report

In [24]:
print("=" * 60)
print("AUTOMATED SALES INTEGRITY REPORT")
print("=" * 60)

print("Total Records       :", len(df))
print("Product Lines       :", df["PRODUCTLINE"].nunique())
print("Outlier Records     :", len(outliers))

print("\nTop Product Lines")
print("-" * 40)

print(summary.head())

print("=" * 60)

AUTOMATED SALES INTEGRITY REPORT
Total Records       : 2823
Product Lines       : 7
Outlier Records     : 81

Top Product Lines
----------------------------------------
PRODUCTLINE
Classic Cars        3919615.66
Vintage Cars        1903150.84
Motorcycles         1166388.34
Trucks And Buses    1127789.84
Planes               975003.57
Name: SALES, dtype: float64


The Automated Sales Integrity & Reporting Engine successfully cleaned and validated the retail sales dataset. The analysis identified 81 revenue outliers and summarized sales across 7 product lines. Classic Cars generated the highest revenue (3,919,615.66), making it the best-performing category. The project demonstrates the use of Python and Pandas for data cleaning, validation, aggregation, and business reporting.